File: 3.1/stamps/stamps.py
"""
ID: giliev91
LANG: PYTHON3
TASK: stamps
"""
from itertools import count


def input():
    with open("stamps.in", "r") as f:
        K, _N = tuple(map(int, f.readline().strip().split()))
        stamps = list(map(int, f.read().replace('\n', '').split()))
        return K, stamps


def solve(K, stamps):
    # s[m] is the minimum number of stamps needed to construct sum m
    # s[m] = min(s[m-n])+1 for each stamp n <= m

    s = [0]
    for m in count(1, 1):
        s.append(0)
        s[m] = 1 + min([s[m-n] for n in stamps if m >= n])

        if s[m] > K:
            return m-1


def output(res):
    with open("stamps.out", "w") as f:
        f.write("%d\n" % res)


K, stamps = input()
res = solve(K, stamps)
output(res)

File: 3.1/stamps/stamps.cpp
/*
ID: giliev91
LANG: C++
TASK: stamps
*/
#include <iostream>
#include <fstream>
using namespace std;
#define MAXM 10000000

int main()
{
    ifstream fin("stamps.in");
    ofstream fout("stamps.out");

    int K, N;
    fin >> K >> N;

    int *stamps = new int[N];
    for (int i = 0; i < N; i++)
        fin >> stamps[i];

    unsigned char *s = new unsigned char[MAXM];
    s[0] = 0;
    int m = 0, min;

    while (s[m++] <= K)
    {
        min = MAXM;
        for (int i = 0; i < N; i++)
            if (m >= stamps[i] && s[m - stamps[i]] < min)
                min = s[m - stamps[i]];

        s[m] = min + 1;
    }

    fout << (m - 2) << "\n";
    return 0;
}